pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -1823,3 +1879,109 @@ func makeListenerKey(l gatev1alpha2.Listener) string {
+	var routeParentStatus *gatev1alpha2.RouteParentStatus
+
+	// Check if we need to update an existing parent reference
+	for i, parent := range routeStatus.Parents {
+		if parent.ControllerName != controllerName {
+			continue
+		}
+
+		parentRef := parent.ParentRef
+
+		if parentRef.Group == nil || *parentRef.Group != gatev1alpha2.GroupName {
+			continue
+		}
+		if parentRef.Kind == nil || *parentRef.Kind != kindGateway {
+			continue
+		}
+		if parentRef.SectionName != nil && *parentRef.SectionName != listener.Name {
+			continue
+		}
+
+		namespace := routeMeta.Namespace
+		if parentRef.Namespace != nil {
+			namespace = string(*parentRef.Namespace)
+		}
+
+		if namespace == gateway.Namespace && string(parentRef.Name) == gateway.Name {
+			log.Ctx(ctx).Debug().Msgf("Found existing parent reference for route %s/%s", routeMeta.Namespace, routeMeta.Name)
+			routeParentStatus = &routeStatus.Parents[i]
+		}
+	}
+
+	// No existing parent reference was found - create a new one
+	if routeParentStatus == nil {
+		log.Ctx(ctx).Debug().Msgf("Adding parent reference for route %s/%s", routeMeta.Namespace, routeMeta.Name)
+
+		group := gatev1alpha2.Group(gatev1alpha2.GroupName)
+		kind := gatev1alpha2.Kind(kindGateway)
+		namespace := gatev1alpha2.Namespace(gateway.Namespace)
+
+		// Only set the section name if the route requests it
+		var sectionName *gatev1alpha2.SectionName
+		for _, p := range routeSpec.ParentRefs {
+			if p.SectionName != nil && *p.SectionName == listener.Name {
+				sectionName = p.SectionName
+			}
+		}
+
+		routeStatus.Parents = append(routeStatus.Parents, gatev1alpha2.RouteParentStatus{
+			ParentRef: gatev1alpha2.ParentRef{
+				Group:       &group,
+				Kind:        &kind,
+				Namespace:   &namespace,
+				Name:        gatev1alpha2.ObjectName(gateway.Name),
+				SectionName: sectionName,
+			},
+			ControllerName: controllerName,
+		})
+
+		routeParentStatus = &routeStatus.Parents[len(routeStatus.Parents)-1]
+	}
+
+	// Don't write duplicate conditions
+	shouldUpdateConditions := true
+	if len(routeParentStatus.Conditions) > 0 {
+		lastCondition := routeParentStatus.Conditions[len(routeParentStatus.Conditions)-1]
+
+		if lastCondition.Type == string(gatev1alpha2.ConditionRouteAccepted) && lastCondition.Status == metav1.ConditionTrue {
+			shouldUpdateConditions = false
+		}
+	}
+
+	if shouldUpdateConditions {
+		log.Ctx(ctx).Debug().Msgf("Route %s/%s was attached to gateway", routeMeta.Namespace, routeMeta.Name)
+
+		routeParentStatus.Conditions = append(routeParentStatus.Conditions, metav1.Condition{
+			Type:               string(gatev1alpha2.ConditionRouteAccepted),
+			Status:             metav1.ConditionTrue,
+			ObservedGeneration: routeMeta.Generation,
+			LastTransitionTime: metav1.NewTime(timeNow()),
+			Reason:             string(gatev1alpha2.ConditionRouteAccepted),
+			Message:            "The route was attached to the Gateway",
+		})
+	}
```

Comment:
In the [`RouteParentStatus` documentation](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.RouteParentStatus), and for the `conditions` field, it is said:
```
If the Route’s ParentRef specifies an existing Gateway that supports Routes of this kind AND that Gateway’s controller has sufficient access, then that Gateway’s controller MUST set the “Accepted” condition on the Route, to indicate whether the route has been accepted or rejected by the Gateway, and why.
```

According to this, I think the condition `ConditionRouteAccepted` should not always be true, it should also inform if the attachment failed and probably give a reason.
But it also says that there are cases where it is not necessary:
```
There are a number of cases where the “Accepted” condition may not be set due to lack of controller visibility, that includes when:

- The Route refers to a non-existent parent.
- The Route is of a type that the controller does not support.
- The Route is in a namespace the controller does not have access to.
```

Here, if I didn’t overlook something, it seems that we would never update the RouteStatus with a status false.
Also, while we should prevent overriding another Gateway/Controller `RouteParentStatus`, I'm not sure we do need to keep a "state" and update it.

Reply:
Good point! I'll have a look if I can add this with minimal blast radius 😅

Reply:
I'm collecting route status conditions now and passing them through to the status maker function. I haven't yet finished the work on that yet, but perhaps you can take a look to see if the approach generally makes sense so far?

Reply:
> not sure we do need to keep a "state" and update it.

Sorry, I completely missed this ☝️  Do you mean that we shouldn't search for existing route parent object when updating the status? I've done it because [RouteStatus](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.RouteStatus) docs have this paragraph:
> When this route attaches to a parent, the controller that manages the parent must add an entry to this list when the controller first sees the route and should update the entry as appropriate when the route or gateway is modified.